python_modules/dagster-graphql/dagster_graphql/schema/roots/query.py

```diff
@@ -416,8 +421,38 @@ def resolve_runConfigSchemaOrError(self, graphene_info, **kwargs):
     def resolve_instance(self, graphene_info):
         return GrapheneInstance(graphene_info.context.instance)
 
-    def resolve_assetNodes(self, graphene_info):
-        return get_asset_nodes(graphene_info)
+    def resolve_assetNodes(self, graphene_info, **kwargs):
+        load_materializations = kwargs.get("loadMaterializations")
+        if "pipeline" in kwargs:
+            sel = RepositorySelector.from_graphql_input(kwargs.get("pipeline"))
+            repo_loc = graphene_info.context.get_repository_location(sel.location_name)
+            repo = repo_loc.get_repository(sel.repository_name)
+            external_asset_nodes = repo.get_external_asset_nodes(
+                kwargs.get("pipeline").get("pipelineName")
+            )
+            asset_nodes = (
+                [GrapheneAssetNode(repo, asset_node) for asset_node in external_asset_nodes]
+                if external_asset_nodes
+                else []
+            )
+        else:
+            asset_nodes = get_asset_nodes(graphene_info)
+
+        asset_keys = set(
+            AssetKey.from_graphql_input(asset_key) for asset_key in kwargs.get("assetKeys", [])
+        )
+        matching = [node for node in asset_nodes if not asset_keys or node.assetKey in asset_keys]
+        if not matching:
+            return []
+
+        if load_materializations:
+            events_by_key = graphene_info.context.instance.get_latest_materialization_events(
+                [node.assetKey for node in matching]
+            )
+            for node in matching:
+                node.attach_latest_materialization(events_by_key.get(node.assetKey), True)
```

Comment:
Strongly prefer to have the graphene objects not be mutable.  Could we get the assetKeys up front before instantiating the nodes?

Reply:
Ahh yeah this is definitely awkward, will fix. I think we can do it beforehand but I wanted to avoid building objects and then rebuilding the same objects a second time.

Reply:
Hey @prha - I addressed this with a new commit here.  It's a bit awkward because `GrapheneAssetNode` requires a repo and external_asset_node, and we can't load the materialization events earlier because we don't know which asset keys we'll want until we load the asset nodes.

I tried creating a set of `[repo, asset_node]` tuples, loading the events, and then building `GrapheneAssetNode` instances, but having a new tuple type seemed a bit silly. I ended up adding getters to `GrapheneAssetNode` so that the `load_materializations` code path can create a new set of GrapheneAssetNodes with the events attached once they're loaded.